### ISSUE TITLE
Smooth Scrolling

### ISSUE THREAD

**Issue by casey** [jwilm/alacritty#2053]
In all terminal emulators that I'm familiar with, new lines of text appear instantly on the bottom of the screen, shifting all previous lines up one line height. Compared with smooth scrolling, e.g. a web page or a PDF, this is jerky and visually unappealing.

For an example of what a smooth scrolling terminal might look like, check out this [potato quality youtube video](https://www.youtube.com/watch?v=Iju_pOQM0a0).

I've thought a bit about how to implement smooth scrolling for a terminal, and it isn't immediately obvious that it's possible to do in a visually appealing and consistent manner. A new line of text in a terminal comes into existence at the moment it is displayed. Content on a web page, on the other hand, exists outside of the viewing rectangle, and smooth scrolling the viewing rectangle down to view additional content is conceptually and technically simple.

A simple implementation might be that new lines are initially off screen, and they scroll into view at a constant speed. However, in this case, if the terminal produces lines of text faster than they can be scrolled into view, the terminal will back up, and the user will be stuck watching stale lines of text scroll by.

An alternative approach would be to define a function which, given the height of the offscreen text, outputs a scroll speed which would be fast enough to scroll the text into view quickly enough that the user isn't annoyed by the amount of time it takes.

Another wrinkle is that I'm unfamiliar with terminal plumbing, and so I don't know if the terminal has enough information to smoothly scroll text in all cases. For example, when a user is in a text editor such as vim, which has a status line at the bottom of the screen, does the terminal "know" that the text region is scrolling?

This is might be a feature that's more trouble than it's worth, but I wanted to open this issue to hear what people thought, and whether it merited further exploration.

**Comment by chrisduerr** [comment]
What is definitely possible and has been discussed before is smooth scrolling while actually scrolling. This is also already done by a few other terminal emulators and it's not completely impossible to get it working in Alacritty, though it would require a bit of effort.

Now when it comes to scrolling in new lines of text using smooth scrolling, I feel like there's an extremely high chance that would actually annoy the living crap out of most users. In theory it might be possible to render new messages below the visible area and then scroll them into view though. But I don't really see Alacritty introducing something like that.

**Comment by casey** [comment]
I think a naïve implementation would likely be incredibly annoying. Perhaps there's a good enough implementation that's an improvement on the status quote though. For example, scrolling very, very fast, maybe over only a few frames, such that the user never needs to wait, and there's still an enhanced perception of smoothness.

**Comment by nixpulvis** [comment]
I personally want nothing to do with smooth scrolling, and will disable this feature if implemented. That said, I could see scrolling the scrollback buffer "smoothly" being nicer for some people. The idea of scrolling in input rather than just rendering it in place seems very wrong to me.

**Comment by casey** [comment]
@nixpulvis Are you assuming that there would be some issue with the implementation, or is the feature itself, even if it was implemented well, unappealing to you?

I definitely dislike complexity, but when I look at smoothly scrolling text it just looks so much nicer to me.

**Comment by nixpulvis** [comment]
Well theoretically any animation like this is going to "cost" something, both at runtime, and in code complexity. That's not to say it couldn't be done well. I can see many users liking smooth scrolling of the terminal history, and I wouldn't even mind if that was the default. I'll still probably leave this off myself, as I'm growing more and more accustomed to knowing my text is lined up all the time. It's sorta a weird thing honestly, but like now when I use other peoples terminals I find myself obsessively scrolling to the bottom (even when I'm already there).

The part that I find harder to imagine being a good feature is smooth scrolling new text coming into the terminal. This just doesn't seem in line with the design of the project, even if it's easy to implement after getting smooth history scrollback.

**Comment by casey** [comment]
> The part that I find harder to imagine being a good feature is smooth scrolling new text coming into the terminal. This just doesn't seem in line with the design of the project, even if it's easy to implement after getting smooth history scrollback.

It's definitely an unknown, since we don't know of any implementations out there. It probably makes sense to start with smooth scrolling scrollback, and then if anyone's interested they can experiment with new line scrolling.

**Comment by VKoskiv** [comment]
One could implement an acceleration curve to the scrolling so it isn’t annoyingly slow when a lot of text is printed at once, but still has that pleasing smooth effect when only a few lines are printed.
A lot of old DEC terminals used to have this sort of scrolling and I’d love to see a terminal emulator replicate it.

**Comment by raycewest** [comment]
I think this would really shine with pagers or text editors where you often scroll a small chunk at a time.

**Comment by KhaimovMR** [comment]
Scrolling a history with the smoothness could be implemented for the mouse (touchpad) scroll, for example, but for the keyboard/stdout scroll it could be just using it's usual old-fashioned style.

**Comment by raycewest** [comment]
I'm actually kind of interested in messing with this, if it's trivial at least.  Is there an existing proof of concept?

**Comment by chrisduerr** [comment]
> if it's trivial at least

It certainly is not.

> Is there an existing proof of concept?

As far as I'm aware, nobody has bothered with it at all yet.

Contributions are always appreciated of course, however for a terminal emulator, especially with Alacritty's architecture, this is an extremely complex issue when the goal is to get a solution that works well enough to be upstreamed into Alacritty.

**Comment by lewisthompson** [comment]
I don't know exactly what "smooth scrolling" means but with my minor tweak discussed in https://github.com/jwilm/alacritty/issues/2869 the scrolling feels very natural to me with both scroll wheel & MBP trackpad.

**Comment by blueForestIcarus** [comment]
Someone has been working on implementing this for kitty, with some proof of concept gifs to show what "smooth" (ie. pixel) scrolling would look like. https://github.com/kovidgoyal/kitty/pull/1454

**Comment by chrisduerr** [comment]
@blueForestIcarus That only seems to implement the trivial case of scrolling the viewport though. What has been suggested in the original issue is more complicated though both would be possible in theory.

**Comment by r4dr3fr4d** [comment]
I think I may try to implement what I'm looking for with st, at some point
in the future.

**Comment by blueForestIcarus** [comment]
@chrisduerr Scrolling in curses-like program like vim isn't really feasible. The terminal doesn't know that vim is scrolling, It just knows vim is writing lines. I suppose alacrity could implement some sort of escape codes which an application like vim could use to define a "scroll region" or alternatively an escape code for scroll in/scroll-out overwrite (similar to how bold/blink/color work), but you would need vim, and every other application, to adopt the standard.

**Comment by chrisduerr** [comment]
> Smooth scrolling in curses-like program like vim isn't really feasible.

That's definitely true, I was more talking about outside the primary screen buffer. There actually are hardware terminals that do this, it's quite neat.

**Comment by chrisduerr** [comment]
Actually a clarification on the last point: It's not true that it's not possible in vim and there is an escape sequence that allows smooth scrolling with scrolling regions. It's called `DECSCLM` and it allows you to set the scrolling mode to either jump scroll or smooth scroll.

**Comment by nicohansi** [comment]
I am new to alacritty and just came across [this video](https://www.youtube.com/watch?v=RuZUPpmXfT0) about the original VT320 terminal (smooth scrolling at [7:08](https://youtu.be/RuZUPpmXfT0?t=428)).
I would love to see this forgotten feature re implemented in a modern terminal emulator. I immediately thought of alacritty because it has graphics acceleration, making it probably the most feasible candidate for this gimmick.
And who knows what software just still natively supports the `DECSCLM` characters. Maybe vi? less? maybe even vim?

I just want to know now and I am not buying a VT320 for that.

**Comment by freetz13** [comment]
> I am new to alacritty and just came across [this video](https://www.youtube.com/watch?v=RuZUPpmXfT0) about the original VT320 terminal (smooth scrolling at [7:08](https://youtu.be/RuZUPpmXfT0?t=428)).
> I would love to see this forgotten feature re implemented in a modern terminal emulator. I immediately thought of alacritty because it has graphics acceleration, making it probably the most feasible candidate for this gimmick.
> And who knows what software just still natively supports the `DECSCLM` characters. Maybe vi? less? maybe even vim?
> 
> I just want to know now and I am not buying a VT320 for that.

Just... Why this feature still not implemented in ALL terminal emulators???' It's so old, and still so good!

**Comment by blueForestIcarus** [comment]
@chrisduerr my bad, didnt know about DECSCLM, although it looks like you can only set the vertical region.

**Comment by chrisduerr** [comment]
Probably important to note that `DECSLM` is more about rate of scrolling that what is now traditionally understood as "smooth scrolling". So I would take my last comment with a big grain of salt (it should still be possible, the escape just might not be appropriate).

**Comment by blueForestIcarus** [comment]
With an application that uses the alternative buffer, how would smooth scrolling regions work.
Lets say you focus on the left vsplit in vim -> vim sets the scroll region to the appropriate rectangle.
But how does the terminal know what to scroll? Normally vim gets the scroll wheel escape sequence on stdin and rewrites the buffer itself.

Perhaps writing to the bottom or top of the scroll region would scroll in the txt.

**Comment by chrisduerr** [comment]
> Normally vim gets the scroll wheel escape sequence on stdin and rewrites the buffer itself.

That's not correct, vim instructs the terminal to scroll the scrolling region when it is scrolling.

But of course not every change that is scrolling is always done through scrolling region rotations. The application might choose for example to rewrite the screen if it would be more efficient than rotating. The new lines coming in also likely aren't rotated in, but instead just manually written after rotation.

This certainly isn't something trivial to implement and it would never work perfectly. Which is likely at least part of the reason why no terminal cares to implement it.

**Comment by zim0369** [comment]
I read this issue before opening #5541 and I don't think both of the issues mention the same thing, but still depends on what is going on in the dev's mind. Anyways I wanted smooth scrolling via keyboard mostly when we do halfpageup/down. It might look something like what [this plugin](https://github.com/yuttie/comfortable-motion.vim) is trying to do. It makes vim scrolling so much better. I hope something like this can be implemented in alacritty. I know some users might say it will cost something and we should stay on the leaner side but my argument is that there are already a lot of leaner and simple terminals with minimal features available out there. I really like it when projects like alacritty, neovide, etc. are approaching gui improvements.

**Comment by baris-inandi** [comment]
Any updates on this? really would like to see smooth scrolling

**Comment by thewebmasterp** [comment]
I'd love to have this feature as well.

**Comment by askfiy** [comment]
This feature makes sense and I am willing to wait for its implementation ..

**Comment by serhez** [comment]
This would be amazing, specially when using vim/nvim!

**Comment by Hubro** [comment]
Implementing smooth scrolling that works with pagers, editors and other utilities that take full control over the terminal is probably impossible.

However, there are 2 situations where smooth scrolling would make a lot of sense and would be pretty simple to implement, at least conceptually:

- While scrolling the history. When scrolling up or down, draw the new contents above or below the current buffer and animate it into place.
- When new text appears. Render this text below the current buffer and animate it into view.

The animation length should be configurable and default to something short like 200ms.

Animation length should never *stack*. If new text appears while an animation is still playing, the current animation should be cancelled and replaced with a new animation that takes [animation length] ms.

There should also be some sane limits on how "far" you can smooth scroll. For example, if you scroll 100 pages backwards in history then leave scrolling mode, it shouldn't animate the 100 page scroll back to the bottom. If the scroll "distance" is more than X (for example 2x window height), Alacritty should instead jump directly to that position, like it does today.

If any text GUI type of application takes control over the terminal, such as Vim, less, htop and so on, any smooth scrolling should be disabled.

---

Neovide is a Neovim GUI that implements smooth scrolling. Since Neovim supports embedded terminals that use "regular" Vim buffers, it gives an idea of how this could look in Alacritty:

https://user-images.githubusercontent.com/597206/173884939-bf815207-a8f4-48d0-ab13-fa3d4be852e3.mp4

Neovide also happens to be written in Rust and has several of the same dependencies as Alacritty. Perhaps something can be learned from how Neovide handles smooth scrolling.

**Comment by chrisduerr** [comment]
> Implementing smooth scrolling that works with pagers, editors and other utilities that take full control over the terminal is probably impossible.

You'd be surprised, there's *hardware* terminals that do this.

**Comment by freetz13** [comment]
> How can a terminal even know what to animate?

Yeah, [how](https://www.youtube.com/watch?v=tSJfzrSA0ec)?

> the terminal would have to be communicating with the application for this to be possible

Yes.

Vim doesn't care about smooth scrolling. That's why I don't like it. Terminal emulators these days lacks many good old hardware terminals features.

**Comment by Hubro** [comment]
> > How can a terminal even know what to animate?
> 
> Yeah, [how](https://www.youtube.com/watch?v=tSJfzrSA0ec)?

Sorry, I didn't think I needed to include all the context again :smile: I mean, how can a terminal even know what to animate when it's displaying a terminal application that *takes control over the terminal*, like Vim, less, top etc. These application don't simply draw new text, they control all the text on the monitor. They can redraw each character as they want. The terminal can't know how to animate any of this.

The video you linked demonstrates the simple scrolling that I was just advocating for in my previous comment, i.e. animating new text, produced by simple stdin/stdout/stderr terminal applications.

> Vim doesn't care about smooth scrolling. That's why I don't like it. Terminal emulators these days lacks many good old hardware terminals features.

Well it's not that Vim doesn't "care", Vim is a terminal application, so there is no way for it to do smooth scrolling, at least not pixel-wise. I have seen plugins that implement smooth scrolling line-by-line.

If you use a GUI for Vim (such as Neovide for Neovim) they are free to implement smooth scrolling, as you see in my video above.

**Comment by freetz13** [comment]
Sorry for repeating that video.

Looks like the only way to have a smooth scrolling in terminal emulators nowadays is to completely rewrite Vim and others. Smells like a revolution?

**Comment by j4james** [comment]
> how can a terminal even know what to animate when it's displaying a terminal application that _takes full control over the terminal_, like Vim, less, top etc.

Typically the application sends a `DECSTBM` sequence to define the scrolling region, so they can designate a fixed header and/or a fixed footer, while having the region in between scrollable. Then if you want to scroll down, you output a linefeed on the last line of the scroll region. If you want to scroll up, you output an `RI` sequence (reverse index) on the top line.

However, to get smooth scrolling in vim, you need to trick it a little, because by default it'll use the `IL` and `DL` operations to scroll, and those don't take the scrolling mode into account. If you set your `TERM` to something like `vt100`, though, that'll force it to use `LF` and `RF` instead. There are probably better ways to achieve the same thing, but that was just the first thing I tried.

The bottom line is that vim is definitely capable of smooth scrolling, but it'll take some tweaking to get it working correctly.

**Comment by Hubro** [comment]
@j4james Interesting, I've never heard about any of that. I wonder why that spec exists though :thinking: What has it ever been used for? It sounds almost like it was designed for implementing smooth scrolling.

In any case, this seems like considerably more complexity and effort than what I consider the MVP, smooth scrolling the history buffer and smooth scrolling in new content.

**Comment by j4james** [comment]
> EDIT: Seems like DECSTBM is just for setting top and bottom margins, so it wouldn't work for Vim. Anybody using Vim almost certainly has a gutter column and/or multiple horizontal window splits.

You can also set the left and right margins with `DECSLRM`, so that shouldn't be an issue. However, I don't know any terminal emulators that support smooth scrolling with left/right margins, so I can't confirm whether that would work in vim. It is theoretically possible though.

**Comment by DADA30000** [comment]
hey, is there any terminal emulator with neovide animations? (smooth cursor and scrolling)

**Comment by leafOfTree** [comment]
The mouse/touchpad scrolling is smooth, so I emulated the scroll event from the keyboard action on my folk:
https://github.com/leafOfTree/alacritty/commit/16bc68d86b1e96442547dba14cb6cc01c

Some automation tools also can send scroll events repeatedly to obtain smooth scrolling system wide.

https://github.com/alacritty/alacritty/assets/13349592/eb850817-f89e-42bf-a7e6-4e51fb162c80
